### PR TITLE
NAS-11305 / 22.02-RC.2 / Fix/refactor reference parsing for image throughout the services

### PR DESCRIPTION
**Pull request by Qubad786**
## This PR contains:
- `normailize_reference` util based-on researching the docker/cli
- Utilize where-ever we're parsing out registry, image, tag/digest information out of reference

I think, the scope of this PR went larger than the ticket assigned as soon as I dig deeper – I found out that `retrieve_image_digest` was also unable to parse out tag information from digest based reference. So, I did these updates in way that they would be re-used easily and provide us with central means to normalize the images/references.

## Examples
```bash
# Example parsing the digest would be like:
❯ midclt call container.image.normalize_reference 'ghcr.io/k8s-at-home/transmission:v3.00@sha256:355f4036c53c782df1957de0e16c63f4298f5b596ae5e621fea8f9ef02dd09e6' | jq
{
  "reference": "ghcr.io/k8s-at-home/transmission:v3.00@sha256:355f4036c53c782df1957de0e16c63f4298f5b596ae5e621fea8f9ef02dd09e6",
  "image": "k8s-at-home/transmission",
  "tag": "sha256:355f4036c53c782df1957de0e16c63f4298f5b596ae5e621fea8f9ef02dd09e6",
  "registry": "ghcr.io",
  "complete_tag": "ghcr.io/k8s-at-home/transmission@sha256:355f4036c53c782df1957de0e16c63f4298f5b596ae5e621fea8f9ef02dd09e6"
}

# Example digest retrieval given: image, tagged image, digested image and tagged+digested image
❯ midclt call container.image.retrieve_image_digest 'redis'
sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546

❯ midclt call container.image.retrieve_image_digest 'redis:latest'
sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546

❯ midclt call container.image.retrieve_image_digest 'redis@sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546'
sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546

❯ midclt call container.image.retrieve_image_digest 'redis:latest@sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546'
sha256:54ee15a0b0d2c661d46b9bfbf55b181f9a4e7ddf8bf693eec5703dac2c0f5546

# Example bulk pull container images
❯ midclt call -job chart.release.pull_container_images trans '{"redeploy": false}'
Status: (none)
Total Progress: [########################################] 100.00%
{"ghcr.io/k8s-at-home/transmission:v3.00@sha256:355f4036c53c782df1957de0e16c63f4298f5b596ae5e621fea8f9ef02dd09e6": "Updated image"}
```

### CODE REVIEW

**Comment by Qubad786** [comment]
@truenas/middleware this is up for review. Thanks!

**Review by themylogin — APPROVED** [review]
Great job right from the start! Welcome to the team!

File: src/middlewared/middlewared/plugins/docker_linux/client.py [hunk 1]
```diff
@@ -58,14 +58,17 @@ async def _get_manifest_response(self, registry, image, tag, headers, mode, rais
...
+                            f"image={image} tag={tag}: {response['error']}")
 
         return response
 
     @private
     async def get_manifest_call_headers(self, registry, image, headers):
         if registry == DEFAULT_DOCKER_REGISTRY:
             headers['Authorization'] = f'Bearer {await self._get_token(image)}'
+        else:
+            headers['Authorization'] = f'Bearer {await self._get_token(f"{registry}/{image}")}'
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/docker_linux/client.py:71]: Would this be required for custom registries ? `_get_token` gets token from `https://auth.docker.io`

File: src/middlewared/middlewared/plugins/docker_linux/utils.py [hunk 2]
```diff
@@ -1,5 +1,60 @@
...
+
+        tag = matches[-1]
+        tag_pos = tagged_image.find(tag)
+        image = tagged_image[:tag_pos - 1].rsplit(':', 1)[0]
+        sep = '@'
+    elif ':' in tagged_image:
+        image, tag = tagged_image.rsplit(':', 1)
+        sep = ':'
+    else:
+        raise CallError(f'Unable to parse the reference: {reference}')
```
Comment (sonicaj) [inline src/middlewared/middlewared/plugins/docker_linux/utils.py:52]: Will we reach this code path ?

**Review by sonicaj — COMMENTED** [review]
Good job!

File: src/middlewared/middlewared/plugins/docker_linux/client.py (hunk 1 above)
Comment (Qubad786) [reply src/middlewared/middlewared/plugins/docker_linux/client.py:71]: I'd say nice catch there!

File: src/middlewared/middlewared/plugins/docker_linux/utils.py (hunk 2 above)
Comment (Qubad786) [reply src/middlewared/middlewared/plugins/docker_linux/utils.py:52]: Hmm, just being super careful – I think, I can just remove it.

File: src/middlewared/middlewared/plugins/docker_linux/client.py (hunk 1 above)
Comment (Qubad786) [reply src/middlewared/middlewared/plugins/docker_linux/client.py:71]: Scope has further increased as I followed through docker auth specs, we need to be able to get auth endpoint and scopes from response headers when we get 401 while fetching manifests: https://docs.docker.com/registry/spec/auth/token/

+ I need to add support for 2 schema versions that we specify in our headers while we request: https://docs.docker.com/registry/spec/manifest-v2-2/

**Review by sonicaj — APPROVED** [review]
Thanks!

**Review by Ornias1993 — APPROVED** [review]
A great solution to a rather confusing issues, awesome work and great clean code!

**Pull request merged**